source/elements/oneDNN/source/data_model/memory/formats.rst

```diff
@@ -0,0 +1,218 @@
+..
+  Copyright 2019-2020 Intel Corporation
+
+.. default-domain:: cpp
+
+.. cpp:namespace:: dnnl::memory
+
+.. |N| replace:: :math:`N`
+.. |C| replace:: :math:`C`
+.. |D| replace:: :math:`D`
+.. |H| replace:: :math:`H`
+.. |W| replace:: :math:`W`
+.. |G| replace:: :math:`G`
+.. |I| replace:: :math:`I`
+.. |O| replace:: :math:`O`
+
+.. |A| replace:: :math:`A`
+.. |M| replace:: :math:`M`
+.. |LDA| replace:: :math:`LDA`
+
+.. |n| replace:: :math:`n`
+.. |c| replace:: :math:`c`
+.. |d| replace:: :math:`d`
+.. |h| replace:: :math:`h`
+.. |w| replace:: :math:`w`
+
+.. |dimensions| replace:: :math:`\operatorname{dimensions}`
+.. |strides| replace:: :math:`\operatorname{strides}`
+.. |offset| replace:: :math:`\operatorname{offset}`
+
+.. _memory_formats-label:
+
+Memory Formats
+--------------
+
+In oneDNN memory format is how a multidimensional tensor is stored in
+1-dimensional linear memory address space. oneDNN specifies two kinds of
+memory formats: *plain* which correspond to traditional multidimensional
+arrays, and *optimized* which are completely opaque.
+
+Plain Memory Formats
+~~~~~~~~~~~~~~~~~~~~
+
+Plain memory formats describe how multidimensional tensors are laid out in
+memory using an array of |dimensions| and an array of |strides| both of which
+have length equal to the rank of the tensor. In oneDNN the order of dimensions
+is fixed and different dimensions can have certain canonical interpretation
+depending on the primitive. For example, for CNN primitives the order for
+activation tensors is :math:`\{N, C, ..., D, H, W\}`, where |N| stands for
+minibatch, |C| stands for channels, and |D|, |H|, and |W| stand for image
+spatial dimensions: depth, height and width respectively. Spatial dimensions
+may be omitted in the order from outermost to innermost; for example, it is
+not possible to omit |H| when |D| is present and it is never possible to omit
+|W|. Canonical interpretation is documented for each primitive. However, this
+means that the |strides| array plays an important role defining the order in
+which different dimension are laid out in memory. Moreover, the |strides| need
+to agree with |dimensions|.
+
+More precisely, let :math:`T` be a tensor of rank :math:`n` and let
+:math:`\sigma` be the permutation of the |strides| array that sorts it, i.e.
+:math:`\operatorname{strides}[i] \geq
+\operatorname{strides}[j]` if :math:`\sigma(i) < \sigma(j)` for all
+:math:`0 \leq i, j < n`. Then the following must hold:
+
+.. math::
+   \operatorname{stides}[i] \geq
+   \operatorname{strides}[j] * \operatorname{dimensions}[j]
+   \text{ if } \sigma(i) < \sigma(j) \text{ for all } 0 \leq i, j < n.
+
+For an element with coordinates :math:`(i_0, \ldots, i_{n-1})` such that
+:math:`0 \leq i_j < \operatorname{dimensions}[j]` for :math:`0 \leq j < n`, its
+offset in memory is computed as:
+
+.. math::
+
+   \operatorname{offset}(i_0, \ldots, i_{n-1})
+   = \sum_{j=0}^{n-1} i_j * \operatorname{strides}[j].
```

Comment:
btw, do we want to mention `offset0`?

Reply:
Makes sense, but not now. Putting this on my TODO list.